Add CI Matrix for various envs

Bonczidai opened the pull request:
Signed-off-by: Bonczidai Levente <levente.bonczidai@hp.com>

Bonczidai commented:
cluster_docker_images_maintenance use  case was failing - fixed it in commit https://github.com/CloudSlang/cloud-slang-content/commit/f398e26a5395656f3fde085340497935b2b68e4e
@tudorlesan @SamMarkowitz please check it

tudorlesan commented:
looks good to me :+1:

SamMarkowitz commented:
:+1:

Bonczidai commented:
the build passes - but only the coreos test suite is activated now,
when I tried to activate the other test suites, I ran into issues which need to be investigated:
- `com.jcraft.jsch.JSchException: Algorithm negotiation fail` (when other test suites like `default` are activated)
- `java.lang.OutOfMemoryError: PermGen space` (`cadvisor` or `jenkins`)

if you have any thoughts on this, please share it - @orius123 @oritstone @meirwah

orius123 commented on circle.yml:
```diff
@@ -12,6 +12,10 @@ general:
+  environment:
+    # "${CIRCLE_PR_NUMBER}" = "" denotes that the build is not part of a fork PR
+    # ignore test suites on node 1 in fork builds
+    ACTIVE_CI_TEST_SUITES_NODE_1: $(if [ "${CIRCLE_PR_NUMBER}" = "" ]; then echo '!default,coreos,cadvisor'; else echo '!default'; fi)
```
can't you use `-z` or something to check if it's empty?
you can also use this syntax instead of echoing
`[ condition ] && var=value_when_true || var=value_when_false`
matter of taste

Bonczidai replied:
not sure if we can use `[ condition ] && var=value_when_true || var=value_when_false` here (however I really like it) since part after `:` expects a result and this expression return nothing

Bonczidai replied:
for some reason `[ -z "${CIRCLE_PR_NUMBER}" ]` does not work ..

orius123 commented on circle.yml:
```diff
@@ -19,29 +23,33 @@ machine:
-     - pip install -r python-lib/requirements.txt -t cslang-builder/lib/Lib
+     - if [ "${CIRCLE_NODE_INDEX}" = "1" ] && [ "${CIRCLE_PR_NUMBER}" = "" ]; then chmod +x ci-env/create_droplets.sh && ci-env/create_droplets.sh; fi:
+          parallel: true
+     - if [ "${CIRCLE_NODE_INDEX}" = "0" ]; then docker run -d -p 49165:8080 jenkins; fi:
```
can't you group it with `&&` for instance?

orius123 commented on circle.yml:
```diff
@@ -19,29 +23,33 @@ machine:
+          parallel: true
   override:
-    - ./cslang-builder/bin/cslang-builder -ts cadvisor,jenkins,consul -cov
+    - case $CIRCLE_NODE_INDEX in 0) ./cslang-builder/bin/cslang-builder -ts default,jenkins,consul -cov ;; 1) ./cslang-builder/bin/cslang-builder -ts ${ACTIVE_CI_TEST_SUITES_NODE_1} -cov ;; esac:
```
in general try to use `\\` and new lines 
it's easier to read.

orius123 commented on circle.yml:
```diff
@@ -19,29 +23,33 @@ machine:
-    - docker ps -a
-#    - docker stop $(docker ps -a -q)
-#    - docker rm $(docker ps -a -q)
+    - if [ "${CIRCLE_NODE_INDEX}" = "1" ] && [ "${CIRCLE_PR_NUMBER}" = "" ]; then chmod +x ci-env/cleanup_env.sh && ci-env/cleanup_env.sh; fi:
```
maybe extract this `[ "${CIRCLE_NODE_INDEX}" = "1" ] && [ "${CIRCLE_PR_NUMBER}" = "" ]` rule to a var

orius123 commented on ci-env/create_droplets.sh:
```diff
@@ -0,0 +1,54 @@
+
+    echo "$COREOS_MACHINE (ID: $DROPLET_ID) droplet creation request accepted"
+  else
+    echo "Problem occurred: $COREOS_MACHINE droplet creation request - status code: $STATUS_CODE"
```
maybe you should `exit 1` here so we won't run the tests for nothing.

orius123 commented on ci-env/create_droplets.sh:
```diff
@@ -0,0 +1,54 @@
+done
+
+# store droplet IDs in a file to be accessible in other script
+echo ${DROPLET_ID_ACC} > "droplets_${CIRCLE_BUILD_NUM}.txt"
```
*styling comment*
you can put this line inside the if statement

orius123 commented on ci-env/wait_for_droplets_and_update_test_inputs.sh:
```diff
@@ -0,0 +1,93 @@
+
+      if [ "${RESPONSE_BODY_JSON}" = "" ]
+      then
+        inc_and_sleep
```
how can we get here?

Bonczidai replied:
we have 3 possibilities when retrieving droplet information:

- reponse body is **empty**: DO is slow, it accepted the request but the droplet is not registered yet (sleep)
-  **active**: do the logic
-  other state like **new** (sleep)

don't think we should break here, we want to sleep and get to the next iteration of the while loop

orius123 commented on ci-env/wait_for_droplets_and_update_test_inputs.sh:
```diff
@@ -0,0 +1,93 @@
+ipv4_list = obj["droplet"]["networks"]["v4"];
+ip = ""
+for ip_obj in ipv4_list:
+  if ip_obj["type"] == "public":
```
you can use a `filter` method here

orius123 commented on ci-env/wait_for_droplets_and_update_test_inputs.sh:
```diff
@@ -0,0 +1,93 @@
+  done
+  if [ "$DROPLET_STATUS" != "active" ]
+  then
+    echo "Droplet($DROPLET_ID) is not active after ${WAITING_TIME}"
```
again `exit 1` here

orius123 commented on ci-env/wait_for_droplets_and_update_test_inputs.sh:
```diff
@@ -0,0 +1,93 @@
+
+# update inputs files to use actual IP addresses
+DROPLET_IP_ARRAY=(${DROPLET_IP_ADDRESS_ACC})
+sed -i "s/<coreos_host>/${DROPLET_IP_ARRAY[0]}/g" test/io/cloudslang/coreos/cluster_docker_images_maintenance.inputs.yaml
```
don't you want to go over all of the content?

orius123 commented on ci-env/wait_for_droplets_and_update_test_inputs.sh:
```diff
@@ -0,0 +1,93 @@
+echo -e "${DO_DROPLET_SSH_PRIVATE_KEY}" > ${SSH_KEY_PATH}
+
+# update inputs files to use actual ssh key
+sed -i "s/<private_key_file>/${SSH_KEY_PATH}/g" test/io/cloudslang/coreos/cluster_docker_images_maintenance.inputs.yaml
```
same here.
lets look this key all over.

orius123 commented on ci-env/cleanup_env.sh:
```diff
@@ -0,0 +1,24 @@
+  then
+    echo "Droplet($DROPLET_ID) deleted successfully"
+  else
+    echo "Problem occurred: destroying droplet($DROPLET_ID) - status code: $STATUS_CODE"
```
`exit 1` here

orius123 commented:
@Bonczidai I reviewed and left some comments.
Great work as always :100:

Bonczidai commented:
we have python code in our script which was tricky to indent since in python
```python
    print "a"
    print "b"
```
gives indentation error.. so here is the scam I applied to the script:
```python
if True:
    print "a"
    print "b"
```
https://github.com/CloudSlang/cloud-slang-content/commit/264f815f91b5316042b9388d0e6f7c5b60053ab9

@orius123 WDYT?

Bonczidai commented:
I think I addressed all the comments:
- added exit codes
- moved echo statement inside if
- used filter in python script
- sed all over
- extracted rule
- indented python scripts
- grouped and aligned commands on machines (tricky here: Yaml keys on multiple lines + bash && )

orius123 commented:
forgot to exit at `wait_for_droplets_and_update_test_inputs`

orius123 commented:
I like what you did with the indentation!

oritstone commented:
@orius123 we also need to merge it to cloudslang-content-0.8 branch
